### PR TITLE
Issue#1036 adding the possibility to use the ReflectiveClassInspector…

### PR DESCRIPTION
**Pull request by oberstrike**
….kt with a custom ClassLoader

### CODE REVIEW

**Comment by Egorand** [comment]
Thanks for your PR @oberstrike! Could you please revert any unrelated formatting changes? This will make reviewing much easier.

File: kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt [hunk 1]
```diff
@@ -46,11 +46,14 @@ import java.lang.reflect.Field
...
 import java.lang.reflect.Modifier
 import java.lang.reflect.Parameter
+import java.net.URLClassLoader
 import java.util.concurrent.ConcurrentHashMap
 import kotlin.LazyThreadSafetyMode.NONE
 
 @KotlinPoetMetadataPreview
-public class ReflectiveClassInspector private constructor() : ClassInspector {
+public class ReflectiveClassInspector private constructor(
+  private val classLoader: URLClassLoader?
```
Comment (Egorand) [inline kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt:55]: Any reason to type this as `URLClassLoader` and not `ClassLoader`? Will a different implementation of `ClassLoader` work here?

Comment (oberstrike) [reply kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt]: @Egorand I changed this to ClassLoader in the following commit

**Comment by oberstrike** [comment]
When will this be put into the official branch? @Egorand

**Comment by Egorand** [comment]
@oberstrike I'll take another pass this week, but before that, as I mentioned, please revert any non-related formatting changes.

File: buildSrc/src/main/java/Dependencies.kt [hunk 2]
```diff
@@ -35,5 +35,6 @@ object deps {
     const val compileTesting = "com.google.testing.compile:compile-testing:0.15"
     const val jimfs = "com.google.jimfs:jimfs:1.1"
     const val ecj = "org.eclipse.jdt.core.compiler:ecj:4.6.1"
+    const val kotlinCompileTesting = "com.github.tschuchortdev:kotlin-compile-testing:1.3.1"
```
[the remark this answers is not in the document]
Comment (oberstrike) [reply buildSrc/src/main/java/Dependencies.kt]: Done!

File: kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt [hunk 3]
```diff
@@ -503,31 +509,36 @@ public class ReflectiveClassInspector private constructor() : ClassInspector {
   public companion object {
     @JvmStatic
     @KotlinPoetMetadataPreview
-    public fun create(): ClassInspector {
-      return ReflectiveClassInspector()
+    public fun create(classLoader: ClassLoader? = null): ClassInspector {
```
[the remark this answers is not in the document]
Comment (oberstrike) [reply kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt:512]: Okay and that means concretely? Then someone would have to adjust the doc.

**Comment by oberstrike** [comment]
> 
> 
> @oberstrike I'll take another pass this week, but before that, as I mentioned, please revert any non-related formatting changes.
I can't undo this, I run spotless but it still remains...

File: kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt [hunk 4]
```diff
@@ -61,7 +63,8 @@ public class ReflectiveClassInspector private constructor() : ClassInspector {
   private fun lookupClass(className: ClassName): Class<*>? {
     return classCache.getOrPut(className) {
       try {
-        Class.forName(className.reflectionName())
+        if (classLoader == null) Class.forName(className.reflectionName())
+        else Class.forName(className.reflectionName(), true, classLoader)
```
Comment (ZacSweers) [inline kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt:67]: Please add braces

File: kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (ZacSweers) [reply kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt:512]: Yes, that's what I said

File: kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt (hunk 4 above)
Comment (ZacSweers) [reply kotlinpoet-classinspector-reflective/src/main/kotlin/com/squareup/kotlinpoet/classinspector/reflective/ReflectiveClassInspector.kt:67]:
```suggestion
        if (classLoader == null) {
          Class.forName(className.reflectionName())
        } else {
          Class.forName(className.reflectionName(), true, classLoader)
        }
```

**Review by ZacSweers — APPROVED** [review]

**Pull request merged**